### PR TITLE
Added Election Term Metric

### PR DESCRIPTION
**Pull request by meetshah777**
This Metric will Publish the Only Election term number

*Tests:*
Docker Container Testing

*1. Output of tmp file*
```
^election_term
{"current_time":1612872530595}
{"Election_Term":18}$
```
*2. Entry in Election term table*
```
sqlite> select * from Election_Term;
18.0|18.0|18.0|18.0
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by yojs — APPROVED** [review]
I have a high level comment -  can we add javadoc comments for public classes and methods ?

**Review by yojs — DISMISSED** [review]

**Review by rguo-aws — COMMENTED** [review]
looks good. I only have one high level comment. Do we also need to include version number as part of this metric, or term number should be sufficient here ?

**Comment by meetshah777** [comment]
> looks good. I only have one high level comment. Do we also need to include version number as part of this metric, or term number should be sufficient here ?

Election Term will be sufficient.

**Comment by meetshah777** [comment]
> javadoc comments

**Comment by meetshah777** [comment]
> I have a high level comment - can we add javadoc comments for public classes and methods ?

ElectionTermCollector class is implementing all the methods of PerformanceAnalyzerMetricsCollector. I think javadoc comment is not required.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (yu-sun-77) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java:2]: nit: can we change year to 2021 for all new files which need to make change to license header?

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollectorTests.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright <2019> Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (yu-sun-77) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollectorTests.java:2]: nit: can we change year to 2021 for all new files which need to make change to license header?

**Review by yu-sun-77 — COMMENTED** [review]
LGTM, just a nitpick for the year in the license header.

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollectorTests.java (hunk 2 above)
Comment (meetshah777) [reply src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollectorTests.java:2]: sure

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java (hunk 1 above)
Comment (meetshah777) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java:2]: sure

**Review by yu-sun-77 — APPROVED** [review]

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    public ElectionTermCollector(){
+        super(SAMPLING_TIME_INTERVAL, "ElectionTermCollector");
+        value = new StringBuilder();
+    }
+
+    @Override
+    public String getMetricsPath(long startTime, String... keysPath) {
+        // throw exception if keysPath.length is not equal to 2
+        // (Keys should be Pending_Task_ID, start/finish OR current, metadata)
```
Comment (amathur1893) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java:44]: Copy paste error

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollectorTests.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        } catch (RuntimeException ex){
+            //- expecting exception...1 value passed; 0 expected
+        }
+    }
+
+    @Test
+    public void testCollectMetrics() {
+        electionTermCollector.collectMetrics(startTimeInMills);
+        String jsonStr = readMetricsInJsonString(1);
+        assertTrue(jsonStr.contains(ElectionTermValue.Constants.ELECTION_TERM_VALUE));
```
Comment (amathur1893) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollectorTests.java:90]: assert the actual value of the term too.

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollectorTests.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Test
+    public void testWithMockClusterService() {
+        ESResources.INSTANCE.setClusterService(mockedClusterService);
+        electionTermCollector.collectMetrics(startTimeInMills);
+        String jsonStr = readMetricsInJsonString(0);
+        assertNull(jsonStr);
+
+        ESResources.INSTANCE.setClusterService(mockedClusterService);
+        when(mockedClusterService.getMasterService()).thenThrow(new RuntimeException());
```
Comment (amathur1893) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollectorTests.java:102]: I don't think this is needed.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+            PerformanceAnalyzerApp.WRITER_METRICS_AGGREGATOR.updateStat(
+                    WriterMetrics.ELECTION_TERM_COLLECTOR_EXECUTION_TIME, "",
+                    System.currentTimeMillis() - mCurrT);
+
+        } catch (Exception ex) {
+            PerformanceAnalyzerApp.ERRORS_AND_EXCEPTIONS_AGGREGATOR.updateStat(
+                    ExceptionsAndErrors.ELECTION_TERM_COLLECTOR_ERROR, "",
+                    System.currentTimeMillis() - mCurrT);
+            LOG.debug("Exception in Collecting ElectionTerm Metrics: {} for startTime {}",
```
Comment (amathur1893) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java:77]: minor - ElectionTerm -> Election term

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        if (keysPath.length != KEYS_PATH_LENGTH) {
+            throw new RuntimeException("keys length should be " + KEYS_PATH_LENGTH);
+        }
+
+        return PerformanceAnalyzerMetrics.generatePath(startTime, PerformanceAnalyzerMetrics.sElectionTermPath);
+    }
+
+    @Override
+    public void collectMetrics(long startTime) {
+        long mCurrT = System.currentTimeMillis();
```
Comment (amathur1893) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ElectionTermCollector.java:54]: Disable the collector by default

**Review by amathur1893 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**